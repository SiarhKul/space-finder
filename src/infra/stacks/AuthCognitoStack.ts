import {CfnOutput, Stack, StackProps} from "aws-cdk-lib";
import {CfnUserPoolGroup, UserPool, UserPoolClient} from "aws-cdk-lib/aws-cognito";
import {Construct} from "constructs";

export class AuthCognitoStack extends Stack {

  public userPool!: UserPool;
  private userPoolClient!: UserPoolClient;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.createUserPool();
    this.createUserPoolClient();
    this.createAdminsGroup()
  }

  private createUserPool() {
    this.userPool = new UserPool(this, 'SpaceUserPool', {
      selfSignUpEnabled: true,
      signInAliases: {
        username: true,
        email: true
      }
    });

    new CfnOutput(this, 'SpaceUserPoolId', {
      value: this.userPool.userPoolId
    })
  }

  private createUserPoolClient(): void {
    this.userPoolClient = this.userPool.addClient('SpaceUserPoolClient', {
      authFlows: {
        adminUserPassword: true,
        custom: true,
        userPassword: true,
        userSrp: true
      }
    });
    new CfnOutput(this, 'SpaceUserPoolClientId', {
      value: this.userPoolClient.userPoolClientId
    })
  }

  private createAdminsGroup(): void {
    new CfnUserPoolGroup(this, "SpaceAdmins", {
      userPoolId: this.userPool.userPoolId,
      groupName: 'admins'
    })
  }
}
