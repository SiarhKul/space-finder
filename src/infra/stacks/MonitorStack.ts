import {Duration, Stack, StackProps} from "aws-cdk-lib";
import {Construct} from "constructs";
import {Alarm, Metric, Unit} from "aws-cdk-lib/aws-cloudwatch";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import {join} from "path";
import {Topic} from "aws-cdk-lib/aws-sns";
import {LambdaSubscription} from "aws-cdk-lib/aws-sns-subscriptions";
import {SnsAction} from "aws-cdk-lib/aws-cloudwatch-actions";

export class MonitorStack extends Stack {

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const webHookLambda = new NodejsFunction(this, 'webHookLambda', {
      runtime: Runtime.NODEJS_18_X,
      handler: 'handler',
      entry: (join(__dirname, '..', '..', 'services', 'monitor', 'handler.ts'))

    })

    const alarmTopic = new Topic(this, 'AlarmTopic', {
      displayName: 'AlarmTopic',
      topicName: 'AlarmTopic'
    })
    alarmTopic.addSubscription(new LambdaSubscription(webHookLambda))

    const spacesApi4xxAlarm = new Alarm(this, 'spacesApi4xxAlarm', {
      metric: new Metric({
        metricName: "4XXError",
        namespace: "AWS/ApiGateway",
        period: Duration.minutes(1),
        statistic: "Sum",
        unit: Unit.COUNT,
        dimensionsMap: {"ApiName": "SpacesApi"}
      }),
      evaluationPeriods: 1,
      threshold: 5,
      alarmName: 'spacesApi4xxAlarm'
    })

    const topicAction = new SnsAction(alarmTopic)
    spacesApi4xxAlarm.addAlarmAction(topicAction)
    spacesApi4xxAlarm.addOkAction(topicAction)

  }
}
