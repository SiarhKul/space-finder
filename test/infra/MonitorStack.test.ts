import {App} from "aws-cdk-lib";
import {MonitorStack} from "../../src/infra/stacks/MonitorStack";
import {Capture, Match, Template} from "aws-cdk-lib/assertions";

describe('Initial test suite', () => {
  let monitorStackTemplate: Template;

  beforeAll(() => {
      const app = new App({
        outdir: 'cdk.out'
      })

      const monitorStack = new MonitorStack(app, 'MonitorStack');
      monitorStackTemplate = Template.fromStack(monitorStack)
    }
  )

  test('Lambda properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
      "Handler": "index.handler",
      "Runtime": "nodejs18.x"
    })
    // arrange step
    // act step
    // assert step
    expect(true).toBeTruthy();
  });
  test('SNS Topic properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic"
    })

    expect(true).toBeTruthy();
  });
  test('SNS Subscriptions', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription',
      Match.objectEquals({
        Endpoint: {
          "Fn::GetAtt": [
            Match.stringLikeRegexp("webHookLambda"),
            "Arn"
          ]
        },
        Protocol: "lambda",
        TopicArn: {
          Ref: Match.stringLikeRegexp("AlarmTopic")
        }
      }))

    expect(true).toBeTruthy();
  });
  test('SNS Subscriptions.JS apporch', () => {
    const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic')
    const snsName = Object.keys(snsTopic)[0]

    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription',
      Match.objectEquals({
        Endpoint: {
          "Fn::GetAtt": [
            Match.stringLikeRegexp("webHookLambda"),
            "Arn"
          ]
        },
        Protocol: "lambda",
        TopicArn: {
          Ref: snsName
        }
      }))
  })
  test('Alarm actions', () => {
    const alarmActionsCapture = new Capture()
    monitorStackTemplate.hasResourceProperties("AWS::CloudWatch::Alarm", {
      AlarmActions: alarmActionsCapture
    })
    // [ { Ref: 'AlarmTopicD01E77F9' } ]
    const alrm = alarmActionsCapture.asArray()

    expect(alrm).toEqual([{
      Ref: expect.stringMatching(/^AlarmTopic/)
    }])
  })

  // test('Monitor stack snapshot',()=>{
  //   expect(monitorStackTemplate.toJSON()).toMatchSnapshot()
  // })

  test('SNS topic snapshot',()=>{
  const topic=  monitorStackTemplate.findResources("AWS::SNS::Topic")

    expect(topic).toMatchSnapshot()
  })
})
