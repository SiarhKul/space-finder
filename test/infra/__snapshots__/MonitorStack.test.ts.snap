// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suite Monitor stack snapshot 1`] = `
{
  "AlarmTopicD01E77F9": {
    "Properties": {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
